### DISCUSSION TITLE
Обновить плейлист и сохранить дату добавления существующих треков

### DISCUSSION THREAD

**Discussion by amd64fox** [Chimildic/goofy#24] [Идея] [answered]
Есть ли такая возможность добавить из API Spotify такую [фишку](https://developer.spotify.com/documentation/web-api/reference/playlists/remove-tracks-playlist/). Можно самый обычный метод  (Removing all occurrences of specific items), он там первый идет. 
 Сейчас если использовать **saveWithReplace** то он обнуляет время добавления некоторых треков которые прошли проверку, что наводит некоторый бардак в обновляемом плейлисте.

**Comment by Chimildic** [comment]
Можно подправить, чтобы у всех дата обнулялась. Чтобы полностью соответствовать Replace (замене). Из-за особенностей запроса, у первой сотни дата может сохраниться. 

Удаление элементов позволит сохранить дату у остальных. Но становится проблемным добавление новых элементов. Из-за непредсказуемости вновь пришедшей сортировки. Для плейлиста с 10 тысячами треков, в худшем случае, это 10 тысяч запросов на перестановку треков. В среднем, меньше, если заморочиться на вычисление. Но не сильно хочется.

Отсюда остается только метод удаления элементов. Но далеко не у всех фильтров есть инверсия, чтобы собрать удаляемые треки. 

Можно конечно в конце вычислить разницу и получить данные тех, что удалились. И добавлять новые через Append метод. Но к чему столько махинаций ради даты?

**Reply by amd64fox** [reply]
Так я  про удаления и имел ввиду. Добавляю через **saveWithAppend**, полностью устраивает. А этот метод (Removing all occurrences of specific items) чисто для удаления, что собственно и нужно. Просто нет альтернативы для **saveWithReplace**, поэтому и предлагаю создать инструмент только для выборочного удаления, не затрагивая соседние треки.

Думаю этот метод удаления всем будет удобнее использовать, если в их скрипте нужно только удалить часть треков и  важна правильная сортировка треков. У меня таких скриптов не один и не два, поэтому и прошу у вас поддержки.
  
Что касается фильтров, для удаления из плейлиста треков мне за глаза хватит **removeTracks**, а там вроде есть инверсия результата.

  Вот насчет **saveWithReplace** тут я не знаю, если людей устраивает такая выдача с разной датой, то можно и так оставить, можно что то вроде голосования провести, мб не все на это внимание обратили.

**Comment by Chimildic** [comment, marked as answer]
@amd64fox попробуйте обновиться [этой бетой](https://github.com/Chimildic/goofy/blob/beta/library.js). И используйте этот метод:
```
    Playlist.saveWithUpdate({
        id: '',
        tracks: tracks,
        toEnd: true,
    });
```
Описание: 
Обновляет треки плейлиста: добавляет новые; удаляет те, которых нет в массиве; сохраняет оригинальную дату добавления треков. Теряется сортировка заданная в массиве. Игнорируются дубликаты по `id`.

Откатиться назад можно [этим кодом](https://github.com/Chimildic/goofy/blob/main/library.js).

**Reply by amd64fox** [reply]
Ухх... спасибо огромное, проверил, треки теперь удаляет как надо, сортировка стоит как мертвая, ну естественно не считая удаленных треков, добавление не тестировал пока... но из описания не совсем понял зачем удалять треки которых нет в массиве, хотя может я не так понял, в любом случаи огромное спасибо. 
 Еще такой вопрос , это войдет в основной код или сразу запихать эту часть в доп. модули ?

**Reply by Chimildic** [reply]
Нужно как и раньше составить массив треков. Без явных инверсий. 
Внутри `saveWithUpdate` будут найдены треки, которые есть в плейлисте, но нет в массиве. Они будут удалены. 
Треки, которые есть в массиве, но нет в плейлисте будут добавлены в начало или конец. 
Сортировка сохраняется только в рамках "блоков": старых и новых треков. Объединить их сортировку нельзя. 

Я попробовал реализовать перестановку для воссоздания сортировки. Но это занимает действительно огромное количество запросов. С лимитами платформы, это нецелесообразно.

 > это войдет в основной код
 
Да, добавлю. Пока отдельно, посмотреть как работает.

**Reply by amd64fox** [reply]
Поигрался сейчас, в общем с удалением все как выше писал, все четко удаляет, вот добавляет уже с нарушением сортировки,  попробовал разные варианты. Но добавлять есть альтернатива saveWithAppend, сейчас написал скрипт сначала добавляю через saveWithAppend, потом удаляю через saveWithUpdate, с такой последовательностью все работает как надо.

**Reply by Chimildic** [reply]
`saveWithUpdate` добавляет треки также, как `saveWithAppend`. Но если эти треки уже есть в плейлисте, повторно не добавятся.

**Reply by amd64fox** [reply]
Странно, у меня при добавлении через `saveWithUpdate` добавляться все треки из массива с новой датой, а те треки что были в плейлисте удаляются. И это не из-за повторов, специально создал пример, что бы повторов не было.

**Reply by Chimildic** [reply]
Добавляет лайки в конец плейлиста. Остальное не трогает. Не знаю что у вас не так.
```js
function debug() {
    let tracks = Source.getPlaylistTracks('', 'id');
    let savedTracks = Source.getSavedTracks();
    Combiner.push(tracks, savedTracks);
    Playlist.saveWithUpdate({
        id: 'id',
        tracks: tracks,
        toEnd: true,
    });
}
```

**Reply by amd64fox** [reply]
А вы можете сделать вот так :
1) Создать в ручную плейлист например 123 и скопировать туда последние 10 любимых треков например.
2) А потом выполнить это код, подставив свое значение id плейлиста 123 в  `saveWithUpdate`




```
function techno() {

let tracks = Source.getPlaylistTracks('Tehno', '37i9dQZF1DX6J5NfMJS675');

Playlist.saveWithUpdate({
        id: '', 
         name: '123',
        tracks: tracks,
        toEnd: true,
       

    });
}
```


В этой плейлисте техно ровно 60 треков, если все правильно работает то должно получиться 70,  но у меня 60.

**Reply by Chimildic** [reply]
Чтобы треки плейлиста сохранились, они должны быть в массиве. 
Иначе алгоритм смотрит на них как требующие удаления.

**Reply by amd64fox** [reply]
Теперь дошло, спасибо
